calyx-frontend/src/attribute.rs

```diff
@@ -1,146 +1,141 @@
+            Attribute::ST_ID => "ST_ID".into(),
+            Attribute::LOOP => "LOOP".into(),
+            Attribute::START => "START".into(),
+            Attribute::END => "END".into(),
+            // Unknown attribute
+            Attribute::Unknown(s) => s,
         }
     }
 }
 
-impl TryFrom<Vec<(Id, u64)>> for Attributes {
-    type Error = calyx_utils::Error;
-
-    fn try_from(v: Vec<(Id, u64)>) -> CalyxResult<Self> {
-        let mut attrs = LinkedHashMap::with_capacity(v.len());
-        for (k, v) in v {
-            if attrs.contains_key(&k) {
-                return Err(Self::Error::malformed_structure(format!(
-                    "Multiple entries for attribute: {k}"
-                )));
-            }
-            attrs.insert(k, v);
-        }
-        Ok(Attributes {
-            attrs,
-            span: GPosIdx::UNKNOWN,
-        })
-    }
-}
-
-impl WithPos for Attributes {
-    fn copy_span(&self) -> GPosIdx {
-        self.span
-    }
-}
-
-/// Structs that can return an [`Attributes`] instance.
-pub trait GetAttributes {
-    /// Returns an [`Attributes`] instance
-    fn get_attributes(&self) -> &Attributes;
-
-    /// Returns a mutable [`Attributes`] instance
-    fn get_mut_attributes(&mut self) -> &mut Attributes;
-}
-
-impl Attributes {
-    /// Add a new attribute
-    pub fn insert<S>(&mut self, key: S, val: u64)
-    where
-        S: Into<Id>,
-    {
-        self.attrs.insert(key.into(), val);
-    }
-
-    /// Get the value associated with an attribute key
-    pub fn get<S>(&self, key: S) -> Option<&u64>
-    where
-        S: Into<Id>,
-    {
-        self.attrs.get(&key.into())
-    }
-
-    /// Check if an attribute key has been set
-    pub fn has<S>(&self, key: S) -> bool
-    where
-        S: Into<Id>,
-    {
-        self.attrs.contains_key(&key.into())
-    }
-
-    /// Returns true if there are no attributes
-    pub fn is_empty(&self) -> bool {
-        self.attrs.is_empty()
-    }
-
-    /// Remove attribute with the name `key`
-    pub fn remove<S>(&mut self, key: S) -> Option<u64>
-    where
-        S: Into<Id>,
-    {
-        self.attrs.remove(&key.into())
-    }
-
-    /// Set the span information
-    pub fn add_span(mut self, span: GPosIdx) -> Self {
-        self.span = span;
-        self
+impl ToString for Attribute {
+    fn to_string(&self) -> String {
+        Id::from(*self).to_string()
     }
 }
 
-impl<S> Index<S> for Attributes
-where
-    S: Into<Id>,
-{
-    type Output = u64;
+impl TryFrom<String> for Attribute {
+    type Error = Error;
 
-    fn index(&self, key: S) -> &u64 {
-        let idx = key.into();
-        self.get(idx)
-            .unwrap_or_else(|| panic!("No key `{}` in attribute map", idx))
-    }
-}
+    fn try_from(s: String) -> CalyxResult<Self> {
+        if DEPRECATED_ATTRIBUTES.contains(&s.as_str()) {
+            return Err(Error::malformed_structure(format!(
+                "Attribute {s} is deprecated"
```

Comment:
We should probably have a message attached to the deprecated attribute message
which explains what should be done instead. For example the static deprecation
might point the user toward the static group construct